crates/wasmtime/src/frame_info.rs

```diff
@@ -208,6 +197,136 @@ impl StoreFrameInfo {
+            // instructions cover `pc`.
+            Err(0) => None,
+
+            // This would be at the `nth` slot, so we're at the `n-1`th slot.
+            Err(n) => Some(n - 1),
+        }
+    }
+}
+
+impl Drop for StoreFrameInfo {
+    fn drop(&mut self) {
+        let mut info = GLOBAL_INFO.lock().unwrap();
+        for end in self.ranges.keys() {
+            info.unregister(*end);
+        }
+    }
+}
+
+/// This is the dual of `StoreFrameInfo` and is stored globally (as the name
+/// implies) rather than simply in one `Store`.
+///
+/// The purpose of this map is to be called from signal handlers to determine
+/// whether a program counter is a wasm trap or not. Specifically macOS has
+/// no contextual information about the thread available, hence the necessity
+/// for global state rather than using thread local state.
+///
+/// This is similar to `StoreFrameInfo` except that it has less information and
+/// supports removal. Any time anything is registered with a `StoreFrameInfo`
+/// it is also automatically registered with the singleton global frame
+/// information. When a `StoreFrameInfo` is destroyed then all of its entries
+/// are removed from the global frame information.
+#[derive(Default)]
+pub(crate) struct GlobalFrameInfo {
+    // The map here behaves the same way as `StoreFrameInfo`.
+    ranges: BTreeMap<usize, GlobalModuleFrameInfo>,
+}
+
+/// This is the equivalent of `ModuleFrameInfo` except has less code and is
+/// stored within `GlobalFrameInfo`.
+struct GlobalModuleFrameInfo {
+    start: usize,
+    functions: Arc<BTreeMap<usize, FunctionInfo>>,
+
+    /// Note that modules can be instantiated in many stores, so the purpose of
+    /// this field is to keep track of how many stores have registered a
+    /// module. Information is only removed from the global store when this
+    /// reference count reaches 0.
+    references: usize,
+}
+
+lazy_static::lazy_static! {
+    static ref GLOBAL_INFO: Mutex<GlobalFrameInfo> = Default::default();
+}
+
+impl GlobalFrameInfo {
+    /// Returns whether the `pc`, according to globally registered information,
+    /// is a wasm trap or not.
+    pub(crate) fn is_wasm_pc(pc: usize) -> bool {
+        let info = GLOBAL_INFO.lock().unwrap();
+        match func(pc, &info.ranges, |i| (i.start, &i.functions)) {
+            Some((_, info)) => info.instr_pos(pc).is_some(),
+            None => false,
+        }
+    }
+
+    /// Registers a new region of code, described by `(start, end)` and with
+    /// the given function information, with the global information.
+    fn register(
+        &mut self,
+        start: usize,
+        end: usize,
+        functions: &Arc<BTreeMap<usize, FunctionInfo>>,
+    ) {
+        // Note that ideally we'd debug_assert that the information previously
+        // stored, if any, matches the `functions` we were given, but for now
+        // we just assume it's the same.
```

Comment:
Can we check that the start is the same? And as a poor man's check for function, that the `functions.size()` match?